### PR TITLE
Runner - Support mpi mode

### PR DESCRIPTION
**Pull request by abuccts**
__Description__

Support mpi mode.

__Major Revisions__

* concate mpirun command
* support mca and env config
* prepare hostfile and update Ansible host pattern

### CODE REVIEW

File: superbench/runner/ansible.py [hunk 1]
```diff
@@ -78,6 +78,20 @@ def run(self, ansible_config, sudo=False):    # pragma: no cover
...
+    def update_mpi_config(self, ansible_config):
+        """Update ansible config for mpi.
+
+        Args:
+            ansible_config (dict): Ansible config dict.
+
+        Returns:
+            dict: Updated Ansible config dict.
+        """
+        if ansible_config['host_pattern'] != 'localhost':
```
Comment (cp5555) [inline superbench/runner/ansible.py:90]: Why is `localhost` different from others? For example, I give 127.0.0.1. It should be the same, correct?

File: superbench/runner/runner.py [hunk 2]
```diff
@@ -48,7 +48,7 @@ def __set_logger(self, filename):
         """
         SuperBenchLogger.add_handler(logger.logger, filename=str(self._output_path / filename))
 
-    def __validate_sb_config(self):
+    def __validate_sb_config(self):    # noqa: C901
```
Comment (cp5555) [inline superbench/runner/runner.py:51]: Why ignore flake8 warning?

File: superbench/runner/templates/ssh_config.j2 [hunk 3]
```diff
@@ -1,5 +1,5 @@
 {% for host in hostvars.values() | map(attribute='ansible_hostname') | sort %}
-Host node{{ loop.index0 }}
+Host node{{ loop.index0 }} {{ host }}
```
Comment (cp5555) [inline superbench/runner/templates/ssh_config.j2:2]: Why add this?

File: superbench/runner/runner.py [hunk 4]
```diff
@@ -211,6 +234,6 @@ def run(self):
                             'proc_rank': proc_rank
                         }) for proc_rank in range(mode.proc_num)
                     )
-                elif mode.name == 'torch.distributed':
+                elif mode.name == 'torch.distributed' or mode.name == 'mpi':
```
Comment (cp5555) [inline superbench/runner/runner.py:237]: Is our mode.name an enum variable? Here is no `else` to handle the failure scenario.

File: superbench/runner/runner.py [hunk 5]
```diff
@@ -122,6 +134,17 @@ def __get_mode_command(self, benchmark_name, mode):
                     'superbench.benchmarks.{name}.parameters.distributed_backend=nccl'
                 ).format(name=benchmark_name),
             )
+        elif mode.name == 'mpi':
+            mode_command = (
+                'mpirun '
+                '--allow-run-as-root -hostfile hostfile -map-by ppr:{proc_num}:numa '
```
Comment (cp5555) [inline superbench/runner/runner.py:140]: What happens if docker doesn't support `-allow-run-as-root`?

Comment (cp5555) [inline superbench/runner/runner.py:140]: What's meaning of `-map-by`? Would you please all comments to describe the option meaning in the code including `--alow-run-as-root`, `-hostfile`, `-map-by`, etc,?

File: superbench/runner/runner.py [hunk 6]
```diff
@@ -122,6 +134,17 @@ def __get_mode_command(self, benchmark_name, mode):
...
+                '{mca_list} {env_list} {command}'
+            ).format(
+                proc_num=mode.proc_num,
+                mca_list=' '.join(f'-mca {k} {v}' for k, v in mode.mca.items()),
+                env_list=' '.join(f'-x {k}={v}' if v else f'-x {k}' for k, v in mode.env.items()),
+                command=exec_command,
+            )
```
Comment (cp5555) [inline superbench/runner/runner.py:147]: If mode_command doesn't belong to any support feature, it will return `exec_command`. Is it our design?

**Review by cp5555 — CHANGES_REQUESTED** [review]

File: superbench/runner/templates/ssh_config.j2 (hunk 3 above)
Comment (abuccts) [reply superbench/runner/templates/ssh_config.j2:2]: updated

File: superbench/runner/runner.py (hunk 4 above)
Comment (abuccts) [reply superbench/runner/runner.py:237]: added warning in else

File: superbench/runner/runner.py (hunk 5 above)
Comment (abuccts) [reply superbench/runner/runner.py]: added

File: superbench/runner/runner.py (hunk 6 above)
Comment (abuccts) [reply superbench/runner/runner.py:147]: it will return original command

File: superbench/runner/runner.py (hunk 2 above)
Comment (abuccts) [reply superbench/runner/runner.py:51]: function too complex, will separate validation part in the future

File: superbench/runner/ansible.py (hunk 1 above)
Comment (cp5555) [reply superbench/runner/ansible.py:90]: As Yifan mentioned, it's ansible's configuration

File: superbench/runner/runner.py (hunk 2 above)
Comment (cp5555) [reply superbench/runner/runner.py:51]: Shall we fix it this time?

**Review by cp5555 — APPROVED** [review]

**Review by guoshzhao — APPROVED** [review]

**Pull request merged**